# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxFontUtils.cpp
# Commit: 3b719cd12933
# Full Hash: 3b719cd1293370d843691b3656b52b04042d5ec4
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2022-03-02 04:31:47
# Regressor Bug: 1756720
# File Overlap Count: 1
# Description:
#   Bug 1756720 - When using legacy MS Symbol fonts, map character codes from the U+00xx range to the PUA range U+F0xx found in the cmap subtables. r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D139724
# ==============================================================================

diff -r 9849c6e79fa3 -r 3b719cd12933 gfx/thebes/gfxFontUtils.cpp
--- a/gfx/thebes/gfxFontUtils.cpp	Tue Mar 01 17:27:27 2022 +0200
+++ b/gfx/thebes/gfxFontUtils.cpp	Tue Mar 01 15:22:56 2022 +0000
@@ -201,7 +201,8 @@
 
 nsresult gfxFontUtils::ReadCMAPTableFormat4(const uint8_t* aBuf,
                                             uint32_t aLength,
-                                            gfxSparseBitSet& aCharacterMap) {
+                                            gfxSparseBitSet& aCharacterMap,
+                                            bool aIsSymbolFont) {
   enum {
     OffsetFormat = 0,
     OffsetLength = 2,
@@ -286,6 +287,20 @@
     }
   }
 
+  if (aIsSymbolFont) {
+    // For fonts with "MS Symbol" encoding, we duplicate character mappings in
+    // the U+F0xx range down to U+00xx codepoints, so as to support fonts such
+    // as Wingdings.
+    // Note that if the font actually has cmap coverage for the U+00xx range
+    // (either duplicating the PUA codepoints or mapping to separate glyphs),
+    // this will not affect it.
+    for (uint32_t c = 0x0020; c <= 0x00ff; ++c) {
+      if (aCharacterMap.test(0xf000 + c)) {
+        aCharacterMap.set(c);
+      }
+    }
+  }
+
   aCharacterMap.Compact();
 
   return NS_OK;
@@ -420,7 +435,8 @@
 uint32_t gfxFontUtils::FindPreferredSubtable(const uint8_t* aBuf,
                                              uint32_t aBufLength,
                                              uint32_t* aTableOffset,
-                                             uint32_t* aUVSTableOffset) {
+                                             uint32_t* aUVSTableOffset,
+                                             bool* aIsSymbolFont) {
   enum {
     OffsetVersion = 0,
     OffsetNumTables = 2,
@@ -445,6 +461,9 @@
   if (aUVSTableOffset) {
     *aUVSTableOffset = 0;
   }
+  if (aIsSymbolFont) {
+    *aIsSymbolFont = false;
+  }
 
   if (!aBuf || aBufLength < SizeOfHeader) {
     // cmap table is missing, or too small to contain header fields!
@@ -479,6 +498,9 @@
     if (isSymbol(platformID, encodingID)) {
       keepFormat = format;
       *aTableOffset = offset;
+      if (aIsSymbolFont) {
+        *aIsSymbolFont = true;
+      }
       break;
     } else if (format == 4 &&
                acceptableFormat4(platformID, encodingID, keepFormat)) {
@@ -509,13 +531,14 @@
                                 gfxSparseBitSet& aCharacterMap,
                                 uint32_t& aUVSOffset) {
   uint32_t offset;
-  uint32_t format =
-      FindPreferredSubtable(aBuf, aBufLength, &offset, &aUVSOffset);
+  bool isSymbolFont;
+  uint32_t format = FindPreferredSubtable(aBuf, aBufLength, &offset,
+                                          &aUVSOffset, &isSymbolFont);
 
   switch (format) {
     case 4:
       return ReadCMAPTableFormat4(aBuf + offset, aBufLength - offset,
-                                  aCharacterMap);
+                                  aCharacterMap, isSymbolFont);
 
     case 10:
       return ReadCMAPTableFormat10(aBuf + offset, aBufLength - offset,
@@ -762,8 +785,9 @@
                                       uint32_t aBufLength, uint32_t aUnicode,
                                       uint32_t aVarSelector) {
   uint32_t offset, uvsOffset;
-  uint32_t format =
-      FindPreferredSubtable(aCmapBuf, aBufLength, &offset, &uvsOffset);
+  bool isSymbolFont;
+  uint32_t format = FindPreferredSubtable(aCmapBuf, aBufLength, &offset,
+                                          &uvsOffset, &isSymbolFont);
 
   uint32_t gid;
   switch (format) {
@@ -772,6 +796,12 @@
                 ? MapCharToGlyphFormat4(aCmapBuf + offset, aBufLength - offset,
                                         char16_t(aUnicode))
                 : 0;
+      if (!gid && isSymbolFont) {
+        if (auto pua = MapLegacySymbolFontCharToPUA(aUnicode)) {
+          gid = MapCharToGlyphFormat4(aCmapBuf + offset, aBufLength - offset,
+                                      pua);
+        }
+      }
       break;
     case 10:
       gid = MapCharToGlyphFormat10(aCmapBuf + offset, aUnicode);